src/main/java/cms/gov/madie/measure/services/TerminologyValidationService.java

```diff
@@ -0,0 +1,139 @@
+package cms.gov.madie.measure.services;
+
+import cms.gov.madie.measure.dto.ValueSetsSearchCriteria;
+import cms.gov.madie.measure.exceptions.InvalidTerminologyException;
+import gov.cms.madie.models.cql.terminology.CqlCode;
+import lombok.RequiredArgsConstructor;
+import lombok.extern.slf4j.Slf4j;
+import org.apache.commons.lang3.StringUtils;
+import org.springframework.boot.json.GsonJsonParser;
+import org.springframework.stereotype.Service;
+import org.springframework.util.CollectionUtils;
+
+import java.util.List;
+import java.util.Map;
+import java.util.stream.Collectors;
+
+@Slf4j
+@Service
+@RequiredArgsConstructor
+public class TerminologyValidationService {
+  private final TerminologyServiceClient terminologyServiceClient;
+
+  public void validateTerminology(String elm, String accessToken) {
+    if (StringUtils.isBlank(elm)) {
+      return;
+    }
+    validateValueSets(elm, accessToken);
+    validateCodes(elm, accessToken);
+  }
+
+  public void validateValueSets(String elm, String accessToken) {
+    List<String> oids = getValueSetOids(elm);
+    if (CollectionUtils.isEmpty(oids)) {
+      return;
+    }
+    var searchCriteria =
+        ValueSetsSearchCriteria.builder()
+            .includeDraft(true)
+            .valueSetParams(
+                oids.stream()
+                    .map(oid -> ValueSetsSearchCriteria.ValueSetParams.builder().oid(oid).build())
+                    .toList())
+            .build();
+    terminologyServiceClient.fetchValueSets(searchCriteria, accessToken);
+  }
+
+  public void validateCodes(String elm, String accessToken) {
+    List<CqlCode> cqlCodes = getCqlCodes(elm);
+    if (CollectionUtils.isEmpty(cqlCodes)) {
+      return;
+    }
+    List<CqlCode> validatedCodes = terminologyServiceClient.validateCodes(cqlCodes, accessToken);
+    // throw error if there is at least one invalid code
+    CqlCode invalidCode =
+        validatedCodes.stream().filter(cqlCode -> !cqlCode.isValid()).findFirst().orElse(null);
+    if (invalidCode != null) {
+      log.error("Invalid cql codes: " + invalidCode.getCodeId());
+      throw new InvalidTerminologyException("CQL Code", invalidCode.getCodeId());
+    }
+  }
+
+  public List<String> getValueSetOids(String elm) {
+    var library = getLibrary(elm);
+    var valueSets = (Map<String, Object>) library.get("valueSets");
+    if (valueSets == null) {
+      return null;
+    }
+    var valueSetDefs = (List<Map<String, String>>) valueSets.get("def");
```

Comment:
I'm wondering if using a strict type instead of type casting would be preferred. I'll have to look it up.